vault/request_handling.go

```diff
@@ -833,7 +833,22 @@ func (c *Core) doRouting(ctx context.Context, req *logical.Request) (*logical.Re
 	// If we're replicating and we get a read-only error from a backend, need to forward to primary
 	resp, err := c.router.Route(ctx, req)
 	if shouldForward(c, resp, err) {
-		return forward(ctx, c, req)
+		fwdResp, fwdErr := forward(ctx, c, req)
+		if fwdErr != nil && err != logical.ErrReadOnly {
```

Comment:
Is `err != logical.ErrReadOnly` necessary considering what `shouldForward` is checking? Also, since the response could contain the error, should that also be checked here and its error added to the `fwdErr`?

Reply:
Thanks for the context. It was very helpful. I don't think `Err.Error() != logical.ErrReadOnly.Error()` is necessary, though maybe you can use `errors.Is(...)` instead. 
As for the second question, since you have added `err` from `c.router.Route(ctx, req)` to `fwdErr`, just note that the error could also be in the `resp` portion as well. What I mean is that in line 834, `err` could be nil, yet `resp.Error()` could contain the context you are looking for. So, in that situation, you are only adding a nil to the `fwdErr`.

Reply:
@hghaf099 I don't think `errors.Is` is correct; note that we want to ignore _exactly_ `ErrReadOnly` and don't want to ignore (i.e., we want to preserve) anything that wraps `ErrReadOnly`: https://go.dev/play/p/fi1cOYCGr3S gives us the wrong result. We don't want to ignore this wrapped `ErrReadOnly` because it could contain context on why `ErrReadOnly` happened.

Reply:
Ah, I see. Thanks again for the quick experiment. Then, the original form is good with me!